### DISCUSSION TITLE
Duplicate Annotations in Core Files

### DISCUSSION THREAD

**Discussion by Sdfendor** [OpenMage/magento-lts#2583] [Questions & Answers]
While updating our shop to the most recent version and then again in PR #2577 we noticed, that in core files the **category** and **package** annotations are duplicated, first in the license docblock and then after that in the class docblock. Isn't that redundant? Couldn't one occurrence be removed? 
I'm asking this questions instead of already working on a PR because we noticed that those duplicates aren't only accidentally present in existing files but are actively added in PRs e.g. the one mentioned above. Maybe there is reason for this duplication that I'm not aware of.

example: 
![image](https://user-images.githubusercontent.com/2728018/189367740-33eedf82-bbf0-497e-8ae5-c87bc76f0af0.png)

**Comment by fballiano** [comment]
I don't know why we have body category and package, probably category is meant to be in a project where there are both openmage and other softwares?

**Reply by Sdfendor** [reply]
You mean having both annotations at all? Or is the body one the one below the license notice?

**Reply by fballiano** [reply]
ah ok, you mean the 2 blocks, I didn't consider them duplicate since the first one is more or less the general copyright for the file, while the second one is relative to the class.

it is true that we'll never have more than a class per file, but that's the reason i think

**Reply by Sdfendor** [reply]
I don't see this eventuality happen in the foreseeable future, maybe never to be honest. The autoloader works on a class per file basis after all. One would have to work around some problems to support multiple classes per file. It's not a big problem or anything to have those duplicate annotations, but I'm not a big fan of redundancy where there is no need for it.

**Reply by fballiano** [reply]
it will never happen for sure, i was just trying to understand why it was done that way since the beginning of time

**Reply by Sdfendor** [reply]
Additionally I got the idea to ask because the last update touched all the files about the "do not change this file to receive updates"-disclaimer. That meant for me there is "no tabu" to change those parts that are present - as you say - since the beginning of time.

**Reply by fballiano** [reply]
there's no tabu, the only thing we have to respect is the copyright part.

i didn't remove any of those docblocks because i thought the reason they existed is the one i was saying before, that's all

**Comment by addison74** [comment]
Good observation.

From the first comment block I would keep the ```@copyright``` and ```@license``` that I would move to the second comment block.

On the ```@author``` line I would remove the email address or change it to a valid one.

On the ```@license``` line I would add the email address ```license@magento.com``` (if it is functional).

The new order should be:
```
@category
@package
@author
@copyright
@license
```

**Reply by Sdfendor** [reply]
In any case such a restructuring change in the two docblocks of each file (or almost each file) would require quite some time or some sort of (semi-)automatic script to help with this.

**Reply by sreichel** [reply]
If we want to change this, we could include it into #2295?

I can test if removal of duplicated annotations effects phpDocumentor/phpDox.

**Reply by Sdfendor** [reply]
To test that would be great!

**Reply by addison74** [reply]
After all, my proposal is not so bad but this depends on the author of that PR @justinbeaty. I would like to know his opinion.

In addition, we must agree on the format of that block. The order proposed by me should be fine, I also requested a change regarding the licensed line. Regarding the copyright, there will be one or two lines as the case may be, the problem solved in #2295.

In PHPStorm the docblock is desplayed only for the first lines, I see OpenMage. it was a very good change to be like that.

**Reply by Sdfendor** [reply]
I for one would be fine with just removal of the duplications. I have no opinion about reordering annotations.

**Reply by sreichel** [reply]
My 3 cents ...

- remove email okay
- remove duplicates when not required okay
- reorder seems unnecessary

**Reply by fballiano** [reply]
since @author doesn't allow anything but an email address, and that email address is dead... let's remove it

**Reply by justinbeaty** [reply]
Just catching up on this. I would say that the `@category` and `@package` could be removed from the first docblock, and kept only in the second one. I think that's what @ADDISON74 was suggesting.

**Reply by fballiano** [reply]
If you want I’ve some preg_replace ready for that

**Reply by sreichel** [reply]
Please wait with this. Its possible that phpdoc/... use this in some way. I'd no time to test ...

**Reply by justinbeaty** [reply]
Personally I think it may be better to do in a separate PR. If we merge it into #2295 then it will make it harder to verify that commit via the git diff command I posted over there.

But I am okay either way, and I can easily rebase @2295 if needed.

**Reply by sreichel** [reply]
Separate PR means we touch nearly every file again :(

My idea was ... if you already have a script for contributors, lets extend it :)

**Reply by justinbeaty** [reply]
> Separate PR means we touch nearly every file again :(
> 
> My idea was ... if you already have a script for contributors, lets extend it :)

We already did touch all the files in #2297 as well as #2562, #2558, etc, so we are already past doing it all in just one commit.

The `update-copyright.php` script wouldn't be the best place to do that, since this is a one-and-done task. But in any case, let's wait to make sure phpdoc isn't using this at all.

**Comment by sreichel** [comment]
My PR just added docblocks in the same way they are used all over openmage - to group classes in phpdocumentor correctly.

I did not test, but having package in both blocks seems not required ... but ... why???

**Reply by Sdfendor** [reply]
Was just a quick example I had saved in the clipboard, nothing especially "bad" with the PR, not at all. I didn't want to suggest anything like that. I simply noticed the duplication and wanted to know why it is present in the whole core alltogether and whether there is a reason for the duplication besides "it was always like that".

**Reply by sreichel** [reply]
> wanted to know why it is present in the whole core alltogether and whether there is a reason for the duplication besides "it was always like that".

Yes, it is present all over core. There are only a few files without 2nd docblock.

Reasons ... not sure about this, but first block belongs to the file itself (copyright, disclaimer ...) and second belongs to the class (package, methods, ...)

Not sure if we have to touch this.

**Reply by fballiano** [reply]
exactly what I thought

**Reply by Sdfendor** [reply]
I'm totally fine with the two docblocks, but I didn't understand the duplicate: 
![image](https://user-images.githubusercontent.com/2728018/189376796-292208ea-269c-45b4-975d-7c8c8e9160cf.png)

I can follow the possible explanation by @fballiano but still I would have thought as you said:
 first block => file => copyright, license,
 second block => class =>category, package, maybe author
